src/sampling.jl

```diff
@@ -42,6 +42,47 @@ function direct_sample!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
 end
 direct_sample!(a::AbstractArray, x::AbstractArray) = direct_sample!(Random.GLOBAL_RNG, a, x)
 
+# check whether we can use T to store indices 1:n exactly
+for T in unique!([Int,UInt,Int64,UInt64,Int128,UInt128])
+    @eval isexact(n, ::Type{$T}) = true
+end
+isexact(n, ::Type{T}) where {T<:Integer} = n ≤ typemax(T)
```

Comment:
AFAICT LLVM is smart enough to optimize the generic method statically, so the loop isn't needed:
```julia
julia> isexact(n, ::Type{T}) where {T<:Integer} = n ≤ typemax(T)
isexact (generic function with 1 method)

julia> @code_llvm isexact(2, Int)
;  @ REPL[20]:1 within `isexact'
define i8 @julia_isexact_1866(i64 signext %0) {
top:
  ret i8 1
}
```